### PR TITLE
New Script: Kimai

### LINKED ISSUE #387: Kimai
https://github.com/community-scripts/ProxmoxVE/discussions/107

Application Name
Kimai OpenSource Timerecording

Website
https://www.kimai.org/de/

Description
Kimai is an open-source time-tracking software designed for freelancers, small teams, and businesses to efficiently track, manage, and analyze work hours. This web-based tool enables users to log their work time with ease, associating entries with specific clients, projects, and tasks.

### PR DESCRIPTION
**Pull request by MickLesk**
## Description

Add new LXC Kimai.

Kimai is an open-source time-tracking software designed for freelancers, small teams, and businesses to efficiently track, manage, and analyze work hours. This web-based tool enables users to log their work time with ease, associating entries with specific clients, projects, and tasks.

**Request**
Fixes https://github.com/community-scripts/ProxmoxVE/issues/387

## Type of change
Please check the relevant option(s):

- [ ] Bug fix (non-breaking change that resolves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (a fix or feature that would cause existing functionality to change unexpectedly)
- [x] New script (a fully functional and thoroughly tested script or set of scripts.)

## Prerequisites
The following efforts must be made for the PR to be considered. Please check when completed:
- [x] Self-review performed (I have reviewed my code, ensuring it follows established patterns and conventions)
- [x] Testing performed (I have tested my changes, ensuring everything works as expected)
- [x] Documentation updated (I have updated any relevant documentation)

## Additional Information (optional)
Some Improvements. (Ignore the systemd error, its my dev branch with some changes)

image

image

### CODE REVIEW

File: ct/kimai.sh [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  / ,<  / / __ `__ \/ __ `/ / 
+ / /| |/ / / / / / / /_/ / /  
+/_/ |_/_/_/ /_/ /_/\__,_/_/   
+                              
+EOF
+}
+header_info
+echo -e "Loading..."
+APP="Kimai"
+var_disk="7"
```
Comment (havardthom) [inline ct/kimai.sh:22]: Having odd number disk size triggers my OCD 🤯

File: ct/kimai.sh [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  cd /opt/kimai
+  composer install --no-dev --optimize-autoloader &>/dev/null
+  bin/console kimai:update &>/dev/null
+  chown -R :www-data .
+  chmod -R g+r .
+  chmod -R g+rw var/
+  sudo chown -R www-data:www-data /opt/kimai
+  sudo chmod -R 755 /opt/kimai
+  echo "${RELEASE}" >/opt/${APP}_version.txt
+  msg_ok "Updated ${APP}"
```
Comment (havardthom) [inline ct/kimai.sh:82]:
```suggestion
  msg_ok "Updated ${APP} to ${RELEASE}"
```

File: ct/kimai.sh [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+function update_script() {
+header_info
+check_container_storage
+check_container_resources
+if [[ ! -d /opt/kimai ]]; then msg_error "No ${APP} Installation Found!"; exit; fi
+RELEASE=$(curl -s https://api.github.com/repos/kimai/kimai/releases/latest | grep "tag_name" | awk '{print substr($2, 2, length($2)-3) }')
+if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_version.txt)" ]]; then
+  msg_info "Stopping Apache2"
+  systemctl stop apache2
+  msg_ok "Services Stopped"
```
Comment (havardthom) [inline ct/kimai.sh:64]:
```suggestion
  msg_ok "Stopped Apache2"
```

File: ct/kimai.sh [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  msg_info "Cleaning Up"
+  rm -rf ${RELEASE}.zip
+  msg_ok "Cleaned"
+  msg_ok "Updated Successfully"
+else
+  msg_ok "No update required. ${APP} is already at ${RELEASE}"
+fi
+exit
+}
+start
```
Comment (havardthom) [inline ct/kimai.sh:97]:
```suggestion

start
```

File: install/kimai-install.sh [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+# https://github.com/tteck/Proxmox/raw/main/LICENSE
+
+source /dev/stdin <<< "$FUNCTIONS_FILE_PATH"
+color
+verb_ip6
+catch_errors
+setting_up_container
+network_check
+update_os
+#add_core_dependencies
```
Comment (havardthom) [inline install/kimai-install.sh:15]:
```suggestion
```

File: install/kimai-install.sh [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  libapache2-mod-php \
+  php8.2-{mbstring,gd,intl,pdo,mysql,tokenizer,zip,xml} 
+msg_ok "Installed Dependencies"
+
+msg_info "Setting up Database"
+DB_NAME=kimai_db
+DB_USER=kimai
+DB_PASS=$(openssl rand -base64 18 | tr -dc 'a-zA-Z0-9' | head -c13)
+MYSQL_VERSION=$(mysql --version | grep -oP 'Distrib \K[0-9]+\.[0-9]+\.[0-9]+')
+sudo mysql -u root -e "CREATE DATABASE $DB_NAME;"
```
Comment (havardthom) [inline install/kimai-install.sh:37]:
```suggestion
mysql -u root -e "CREATE DATABASE $DB_NAME;"
```
no need when root user

File: install/kimai-install.sh [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+sudo mysql -u root -e "CREATE USER '$DB_USER'@'localhost' IDENTIFIED WITH mysql_native_password AS PASSWORD('$DB_PASS');"
```
Comment (havardthom) [inline install/kimai-install.sh:38]:
```suggestion
mysql -u root -e "CREATE USER '$DB_USER'@'localhost' IDENTIFIED WITH mysql_native_password AS PASSWORD('$DB_PASS');"
```

File: install/kimai-install.sh [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+sudo mysql -u root -e "GRANT ALL ON $DB_NAME.* TO '$DB_USER'@'localhost'; FLUSH PRIVILEGES;"
```
Comment (havardthom) [inline install/kimai-install.sh:39]:
```suggestion
mysql -u root -e "GRANT ALL ON $DB_NAME.* TO '$DB_USER'@'localhost'; FLUSH PRIVILEGES;"
```

File: install/kimai-install.sh [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  apache2 \
+  git \
+  expect \
+  composer \
+  mariadb-server \
+  libapache2-mod-php \
+  php8.2-{mbstring,gd,intl,pdo,mysql,tokenizer,zip,xml} 
+msg_ok "Installed Dependencies"
+
+msg_info "Setting up Database"
```
Comment (havardthom) [inline install/kimai-install.sh:32]:
```suggestion
msg_info "Setting up database"
```

File: install/kimai-install.sh [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+sudo mysql -u root -e "GRANT ALL ON $DB_NAME.* TO '$DB_USER'@'localhost'; FLUSH PRIVILEGES;"
+{
+    echo "Kimai-Credentials"
+    echo "Kimai Database User: $DB_USER"
+    echo "Kimai Database Password: $DB_PASS"
+    echo "Kimai Database Name: $DB_NAME"
+} >> ~/kimai.creds
+msg_ok "Set up database"
+
+msg_info "Setup Kimai (Patience)"
```
Comment (havardthom) [inline install/kimai-install.sh:48]:
```suggestion
msg_info "Installing Kimai (Patience)"
```

File: install/kimai-install.sh [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+  
+    ErrorLog /var/log/apache2/error.log
+    CustomLog /var/log/apache2/access.log combined
+
+</VirtualHost>
+EOF
+$STD a2ensite kimai.conf
+$STD a2dissite 000-default.conf  
+$STD systemctl reload apache2
+msg_ok "Created Services"
```
Comment (havardthom) [inline install/kimai-install.sh:99]:
```suggestion
msg_ok "Created Service"
```

File: json/kimai.json [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  "logo": "https://raw.githubusercontent.com/kimai/images/refs/heads/main/logo-transparent.png",
+  "description": "Kimai is an open-source time-tracking software designed for freelancers, small teams, and businesses to efficiently track, manage, and analyze work hours. This web-based tool enables users to log their work time with ease, associating entries with specific clients, projects, and tasks.",
+  "install_methods": [
+    {
+      "type": "default",
+      "script": "/ct/kimai.sh",
+      "resources": {
+        "cpu": 2,
+        "ram": 2048,
+        "hdd": null,
```
Comment (havardthom) [inline json/kimai.json:23]:
```suggestion
        "hdd": "7",
```
🤯

**Review by havardthom — DISMISSED** [review]

File: ct/kimai.sh (hunk 2 above)
Comment (MickLesk) [reply ct/kimai.sh:82]: Shall we make this the default? Currently every LXC is different 😅 I had it like this in the first draft, but thought it looked stupid and removed it again

File: install/kimai-install.sh (hunk 5 above)
Comment (MickLesk) [reply install/kimai-install.sh:15]: my fault, its from my dev branch, removed.

**Comment by MickLesk** [comment]
Tested and work fine @havardthom Ready for Approve

File: json/kimai.json [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  "website": "https://www.kimai.org/",
+  "logo": "https://raw.githubusercontent.com/kimai/images/refs/heads/main/logo-transparent.png",
+  "description": "Kimai is an open-source time-tracking software designed for freelancers, small teams, and businesses to efficiently track, manage, and analyze work hours. This web-based tool enables users to log their work time with ease, associating entries with specific clients, projects, and tasks.",
+  "install_methods": [
+    {
+      "type": "default",
+      "script": "/ct/kimai.sh",
+      "resources": {
+        "cpu": 2,
+        "ram": 2048,
```
Comment (havardthom) [inline json/kimai.json:22]:
```suggestion
        "ram": "2048",
```

File: json/kimai.json [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+  "documentation": "https://www.kimai.org/documentation/",
+  "website": "https://www.kimai.org/",
+  "logo": "https://raw.githubusercontent.com/kimai/images/refs/heads/main/logo-transparent.png",
+  "description": "Kimai is an open-source time-tracking software designed for freelancers, small teams, and businesses to efficiently track, manage, and analyze work hours. This web-based tool enables users to log their work time with ease, associating entries with specific clients, projects, and tasks.",
+  "install_methods": [
+    {
+      "type": "default",
+      "script": "/ct/kimai.sh",
+      "resources": {
+        "cpu": 2,
```
Comment (havardthom) [inline json/kimai.json:21]:
```suggestion
        "cpu": "2",
```

**Review by havardthom — DISMISSED** [review]

File: ct/kimai.sh (hunk 2 above)
Comment (havardthom) [reply ct/kimai.sh]: Yeah I think we should include release version in `msg_ok` where possible

**Review by havardthom — APPROVED** [review]

File: ct/kimai.sh [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+  msg_ok "Started Apache2"
+
+  msg_info "Cleaning Up"
+  rm -rf ${RELEASE}.zip
+  msg_ok "Cleaned"
+  msg_ok "Updated Successfully"
+else
+  msg_ok "No update required. ${APP} is already at ${RELEASE}"
+fi
+exit
```
Comment (judeibe) [inline ct/kimai.sh:95]: Increase the Indent level

**Review by judeibe — DISMISSED** [review]

File: ct/kimai.sh (hunk 15 above)
Comment (MickLesk) [reply ct/kimai.sh:95]: ??? i dont see any problem. same spaces like all other latest ct-scripts

Comment (MickLesk) [reply ct/kimai.sh:95]: 2 spaces, or do you have another example? checked 6 Files and all the same

Comment (MickLesk) [reply ct/kimai.sh:95]: give me an example. (1-2 ct's) i still don't understand it. every current script looks like this ^^

Comment (havardthom) [reply ct/kimai.sh:95]: You are right, every script is like this currently, so it's fine. We should probably change that for all scripts later though, as indenting functions and conditions is common practice for readability

**Review by newzealandpaul — APPROVED** [review]

**Pull request merged**